Nightly build compilation errors in TestSIMD.cpp: Kokkos_Complex.hpp(47): error: static assertion failed with "Kokkos::complex can only be instantiated for a cv-unqualified floating point type"

**Describe the bug**
Nightly builds (all Trilinos builds, standalone Kokkos with cuda/11.2, rocm/5.2.0, intel/2021.3, clang/10...) are failing to compile following merge of these commits:

Ensure that complex is only instantiated for cv-unqualified floating-point type ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/173/changes#113055e3c0b46d9d83f461f8aaa9a6819e28827e) / [githubweb](https://github.com/kokkos/kokkos/commit/113055e3c0b46d9d83f461f8aaa9a6819e28827e))
    SYCL: Use in-order queues in InterOp tests (#6246) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/173/changes#6a8488d78bcec6448bdf42b0258d076de155bd4c) / [githubweb](https://github.com/kokkos/kokkos/commit/6a8488d78bcec6448bdf42b0258d076de155bd4c))
    std_algos: fix wrong corner case for `is_partitioned` (#6257) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/173/changes#07a0f0c2c8ff3ab1a8134e44aeb75c3522854197) / [githubweb](https://github.com/kokkos/kokkos/commit/07a0f0c2c8ff3ab1a8134e44aeb75c3522854197))
    SIMD: Add abs() for all int types (#6069) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/173/changes#35c1e3c37402ff8505410456421145e0cd50812c) / [githubweb](https://github.com/kokkos/kokkos/commit/35c1e3c37402ff8505410456421145e0cd50812c))

**Error message snip**:
```
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/core/src/Kokkos_Complex.hpp(47): error: static assertion failed with "Kokkos::complex can only be instantiated for a cv-unqualified floating point type"
19:27:37           detected during:
19:27:37             instantiation of class "Kokkos::complex<RealType> [with RealType=int]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(327): here
19:27:37             instantiation of "auto absolutes::on_host_serial(const T &) const [with T=int]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(196): here
19:27:37             instantiation of "void host_check_math_op_one_loader<Abi,Loader,UnaryOp,T>(UnaryOp, std::size_t, const T *) [with Abi=Kokkos::Experimental::simd_abi::scalar, Loader=load_element_aligned, UnaryOp=absolutes, T=int32_t]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(256): here
19:27:37             instantiation of "void host_check_math_op_all_loaders<Abi,Op,T...>(Op, std::size_t, const T *...) [with Abi=Kokkos::Experimental::simd_abi::scalar, Op=absolutes, T=<int32_t>]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(352): here
19:27:37             instantiation of "void host_check_all_math_ops<Abi,DataType,n>(const DataType (&)[n], const DataType (&)[n]) [with Abi=Kokkos::Experimental::simd_abi::scalar, DataType=int32_t, n=11UL]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(371): here
19:27:37             instantiation of "void host_check_math_ops<Abi,DataType>() [with Abi=Kokkos::Experimental::simd_abi::scalar, DataType=int32_t]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(553): here
19:27:37             instantiation of "void host_check_math_ops_all_types<Abi,DataTypes...>(Kokkos::Experimental::Impl::data_types<DataTypes...>) [with Abi=Kokkos::Experimental::simd_abi::scalar, DataTypes=<int32_t, uint32_t, int64_t, uint64_t, double>]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(560): here
19:27:37             instantiation of "void host_check_abi<Abi>() [with Abi=Kokkos::Experimental::simd_abi::scalar]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(592): here
19:27:37             instantiation of "void host_check_abis(Kokkos::Experimental::Impl::abi_set<FirstAbi, RestAbis...>) [with FirstAbi=Kokkos::Experimental::simd_abi::scalar, RestAbis=<>]" 
19:27:37 /home/jenkins/weaver/workspace/Kokkos_develop_weaver_cuda_openmp_cuda1122_gcc830_volta-debug/kokkos/simd/unit_tests/TestSIMD.cpp(604): here
```

The static_assert indicates this follows merge of #6251 , @dalg24 any thoughts?

**Please also include the following items to support reproducing the bug**
**Reproducer** (Weaver rhel8 queue, Volta70+Power9):
```bash
module load cuda/11.2.2/gcc/8.3.1  cmake/3.23.1

cmake -DCMAKE_CXX_COMPILER=$KOKKOS_PATH/bin/nvcc_wrapper -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ARCH_POWER9=ON -DKokkos_ARCH_VOLTA70=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=OFF -DKokkos_ENABLE_CUDA_LAMBDA=ON -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_CXX_STANDARD=17 -DCMAKE_BUILD_TYPE=DEBUG -DKokkos_ENABLE_DEBUG=ON -DKokkos_ENABLE_DEPRECATION_WARNINGS=OFF -DKokkos_ENABLE_DEPRECATED_CODE_4=OFF $KOKKOS_PATH

```

Here is a snip from a Hip rocm/5.2 compilation error beginning after list of commits above, but lists `Kokkos::abs` in the error output so maybe this follows from #6069 ? 

```
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/core/src/Kokkos_Complex.hpp:47:3: error: static_assert failed due to requirement 'std::is_floating_point_v<long>' "Kokkos::complex can only be instantiated for a cv-unqualified floating point type"
20:33:05   static_assert(std::is_floating_point_v<RealType> &&
20:33:05   ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:327:29: note: in instantiation of template class 'Kokkos::complex<long>' requested here
20:33:05       return Kokkos::abs<T>(a);
20:33:05                             ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:327:14: note: while substituting deduced template arguments into function template 'abs' [with RealType = long]
20:33:05       return Kokkos::abs<T>(a);
20:33:05              ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:196:40: note: in instantiation of function template specialization 'absolutes::on_host_serial<long>' requested here
20:33:05       expected_result[lane] = unary_op.on_host_serial(arg[lane]);
20:33:05                                        ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:256:3: note: in instantiation of function template specialization 'host_check_math_op_one_loader<Kokkos::Experimental::simd_abi::scalar, load_element_aligned, absolutes, long>' requested here
20:33:05   host_check_math_op_one_loader<Abi, load_element_aligned>(op, n, args...);
20:33:05   ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:352:3: note: in instantiation of function template specialization 'host_check_math_op_all_loaders<Kokkos::Experimental::simd_abi::scalar, absolutes, long>' requested here
20:33:05   host_check_math_op_all_loaders<Abi>(absolutes(), n, first_args);
20:33:05   ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:371:5: note: in instantiation of function template specialization 'host_check_all_math_ops<Kokkos::Experimental::simd_abi::scalar, long, 11UL>' requested here
20:33:05     host_check_all_math_ops<Abi>(first_args, second_args);
20:33:05     ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:553:4: note: in instantiation of function template specialization 'host_check_math_ops<Kokkos::Experimental::simd_abi::scalar, long>' requested here
20:33:05   (host_check_math_ops<Abi, DataTypes>(), ...);
20:33:05    ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:560:3: note: in instantiation of function template specialization 'host_check_math_ops_all_types<Kokkos::Experimental::simd_abi::scalar, int, unsigned int, long, unsigned long, double>' requested here
20:33:05   host_check_math_ops_all_types<Abi>(DataTypes());
20:33:05   ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:592:3: note: in instantiation of function template specialization 'host_check_abi<Kokkos::Experimental::simd_abi::scalar>' requested here
20:33:05   host_check_abi<FirstAbi>();
20:33:05   ^
20:33:05 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi100_rocm520_hip_openmp/kokkos/simd/unit_tests/TestSIMD.cpp:604:3: note: in instantiation of function template specialization 'host_check_abis<Kokkos::Experimental::simd_abi::scalar>' requested here
20:33:05   host_check_abis(Kokkos::Experimental::Impl::host_abi_set());
```

--- a/simd/unit_tests/TestSIMD.cpp
+++ b/simd/unit_tests/TestSIMD.cpp
@@ -316,6 +316,14 @@ class divides {
 };
 
 class absolutes {
+  template <typename T>
+  static KOKKOS_FUNCTION auto abs_impl(T const& x) {
+    if constexpr (std::is_signed_v<T>) {
+      return Kokkos::abs(x);
+    }
+    return x;
+  }
+
  public:
   template <typename T>
   auto on_host(T const& a) const {
@@ -323,10 +331,7 @@ class absolutes {
   }
   template <typename T>
   auto on_host_serial(T const& a) const {
-    if constexpr (std::is_signed_v<T>) {
-      return Kokkos::abs<T>(a);
-    }
-    return a;
+    return abs_impl(a);
   }
   template <typename T>
   KOKKOS_INLINE_FUNCTION auto on_device(T const& a) const {
@@ -334,7 +339,7 @@ class absolutes {
   }
   template <typename T>
   KOKKOS_INLINE_FUNCTION auto on_device_serial(T const& a) const {
-    return Kokkos::abs<T>(a);
+    return abs_impl(a);
   }
 };
 

